File: server/r1cord_server/render/themes.py
"""The 12 MD DOCS themes vendored in ./themes, with metadata parsed exactly as MD DOCS does.

MD DOCS (src/main.js, `themes:read-all`) reads `@name`, `@background`, `@text` and `@layout` from
the theme's header comment. A theme gets the contents-sidebar ("toc") shell when it declares
`@layout toc` or its id starts with `altuit-toc` (the family predates the flag).
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from functools import cache
from pathlib import Path

THEMES_DIR = Path(__file__).parent / "themes"
DEFAULT_THEME_ID = "toolmaker-noir"

# Same patterns as MD DOCS; JS `\w` is ASCII-only, hence re.ASCII.
_NAME = re.compile(r"@name\s+(.+)")
_BACKGROUND = re.compile(r"@background\s+(#[0-9A-Fa-f]{6})")
_TEXT = re.compile(r"@text\s+(#[0-9A-Fa-f]{6})")
_LAYOUT = re.compile(r"@layout\s+(\w+)", re.ASCII)


@dataclass(frozen=True)
class Theme:
    id: str
    name: str
    background: str
    text: str
    layout: str
    css: str

    @property
    def light(self) -> bool:
        """MD DOCS' export default: a theme whose @background red channel is >= 128 opens light."""
        return int(self.background[1:3], 16) >= 128


def parse_theme(theme_id: str, css: str) -> Theme:
    name = _NAME.search(css)
    background = _BACKGROUND.search(css)
    text = _TEXT.search(css)
    layout = _LAYOUT.search(css)
    declared = layout.group(1).strip().lower() if layout else "standard"
    return Theme(
        id=theme_id,
        name=name.group(1).strip() if name else "Untitled Theme",
        background=background.group(1) if background else "#FFFFFF",
        text=text.group(1) if text else "#000000",
        layout="toc" if declared == "toc" or theme_id.startswith("altuit-toc") else "standard",
        css=css,
    )


@cache
def _themes() -> tuple[Theme, ...]:
    # Bytes, not read_text(): the CSS is published verbatim, so no newline translation.
    themes = [parse_theme(path.stem, path.read_bytes().decode("utf-8")) for path in THEMES_DIR.glob("*.css")]
    return tuple(sorted(themes, key=lambda theme: (theme.name.casefold(), theme.id)))


def list_themes() -> list[Theme]:
    return list(_themes())


def get_theme(theme: str) -> Theme:
    """Theme by id or name, case-insensitive. ValueError when unknown."""
    wanted = str(theme).strip().casefold()
    for candidate in _themes():
        if wanted in (candidate.id.casefold(), candidate.name.casefold()):
            return candidate
    raise ValueError(f"unknown theme: {theme!r}")

File: desktop/src/core/render/themes.js
// The 12 MD DOCS themes vendored in r1cord_server/render/themes, with metadata
// parsed exactly as MD DOCS does (src/main.js, themes:read-all).

const fs = require('node:fs');
const path = require('node:path');
const { ValueError } = require('../errors');

const THEMES_DIR = path.join(__dirname, '..', '..', '..', 'r1cord_server', 'render', 'themes');
const DEFAULT_THEME_ID = 'toolmaker-noir';

// Same patterns as MD DOCS; JS \w without the unicode flag is ASCII-only.
const NAME = /@name\s+(.+)/;
const BACKGROUND = /@background\s+(#[0-9A-Fa-f]{6})/;
const TEXT = /@text\s+(#[0-9A-Fa-f]{6})/;
const LAYOUT = /@layout\s+(\w+)/;

class Theme {
  constructor({ id, name, background, text, layout, css }) {
    this.id = id;
    this.name = name;
    this.background = background;
    this.text = text;
    this.layout = layout;
    this.css = css;
  }

  // MD DOCS' export default: a theme whose @background red channel is >= 128 opens light.
  get light() {
    return parseInt(this.background.slice(1, 3), 16) >= 128;
  }
}

function parseTheme(themeId, css) {
  const name = css.match(NAME);
  const background = css.match(BACKGROUND);
  const text = css.match(TEXT);
  const layout = css.match(LAYOUT);
  const declared = layout ? layout[1].trim().toLowerCase() : 'standard';
  return new Theme({
    id: themeId,
    name: name ? name[1].trim() : 'Untitled Theme',
    background: background ? background[1] : '#FFFFFF',
    text: text ? text[1] : '#000000',
    layout: declared === 'toc' || themeId.startsWith('altuit-toc') ? 'toc' : 'standard',
    css,
  });
}

function pyRepr(value) {
  if (typeof value === 'string') return `'${value.replace(/\\/g, '\\\\').replace(/'/g, "\\'")}'`;
  return String(value);
}

let cachedThemes = null;

function themes() {
  if (cachedThemes) return cachedThemes;
  const found = [];
  for (const name of fs.readdirSync(THEMES_DIR)) {
    if (!name.endsWith('.css')) continue;
    const css = fs.readFileSync(path.join(THEMES_DIR, name));
    found.push(parseTheme(path.parse(name).name, css.toString('utf8')));
  }
  found.sort((a, b) => {
    const an = a.name.toLowerCase();
    const bn = b.name.toLowerCase();
    if (an < bn) return -1;
    if (an > bn) return 1;
    if (a.id < b.id) return -1;
    if (a.id > b.id) return 1;
    return 0;
  });
  cachedThemes = found;
  return cachedThemes;
}

function listThemes() {
  return themes().slice();
}

function getTheme(theme) {
  const wanted = String(theme).trim().toLowerCase();
  for (const candidate of themes()) {
    if (wanted === candidate.id.toLowerCase() || wanted === candidate.name.toLowerCase()) {
      return candidate;
    }
  }
  throw new ValueError(`unknown theme: ${pyRepr(theme)}`);
}

module.exports = {
  THEMES_DIR,
  DEFAULT_THEME_ID,
  Theme,
  parseTheme,
  listThemes,
  getTheme,
};
